### ISSUE TITLE
Android: Replace confirmation dialog with undobar when marking folder as read

### ISSUE THREAD

**Issue by manderson23** [samuelclay/NewsBlur#584]
Inspired by https://getsatisfaction.com/newsblur/topics/any-plans-for-undo-action-button-for-incorrectly-mark-all-as-read-on-a-feed-or-folder I thought it might be nice to replace the dialog with an undobar (like GMail/Pocket see https://plus.google.com/+RomanNurik/posts/RA9WEEGWYp6).

It wouldn't interrupt the flow of using the app as much and seems to be an Android UI pattern that is becoming more popular.

Any thoughts?

**Comment by manderson23** [comment]
FYI I still think this would be a nice improvement but had held off thinking about looking into it while the whole SQL vs Realm debate was ongoing.

If we are still considering Realm (even in 6 month + timeframe) it is questionable whether it would be worth investing any time in the back end changes required to support undo. Any refactoring/redesign for Realm could perhaps be performed with consideration to supporting undo.

Thoughts?

**Comment by dosiecki** [comment]
I think this is roughly do-able regarless of DB solution, now.  Either way, we will have to build the same components:

1) a delay system for remote actions
2) a journal-ish system for local actions so they can be un-done
3) the UI

The most difficult part is the second.  It is insufficient to say "undo mark-read folder Foo" to the DB, because the DB doesn't know what stories were unread before the mark.  Rather, we have to temporarily store all the hashes that were marked by the action, so they can be locally un-done when the remote action is cancelled from the delay queue.  Alternatively, we could make the 'read' status of stories a tombstone-like ternary flag, rather than a binary, and require a confirmation write in order to flip them permanantly.

Part 1 is mostly done, as the offline system was built with delayed actions in mind.  The UI is as much work as we want to make it.

All told, it is a *lot* of moving parts.  I would personally triage anything that increases complexity of the app this much a low-to-lowest priority, at least until we are caught up on bugs.  Do we have feedback from users that the confirmation dialogue is causing problems?

**Comment by manderson23** [comment]
As with most of the work I have contributed this is pretty much something I'd like to see. I feel that the confirmation dialog just interrupts the flow when using the app. When I press to mark a folder read I'm pretty sure I want to mark it read. No other action to mark multiple items read (e.g. on a feed) has a confirmation though obviously the folder action is going to effect most stories (apart from perhaps older/newer on a folder). It is also quite jarring UI compared to other apps where the undo action is available.

Undo could obviously be applied to any action that marks multiple items as read and gives a way to back out any accidental mark read actions.

**Comment by dosiecki** [comment]
I noticed that undo has received a bit of attention on getsatisfaction lately and had a thought today:

Implementing this on Android would be much easier if we made one small concession.  If the *localy-visible rollback of the mark-read was delayed* until the next feed metadata refresh, we could eliminate most of the nasty bookkeeping that makes this complex.  We could even trigger the refresh in a lightweight manner so the delay would be small, provided the user is online.

I personally think this would be an acceptable compromise given that the downside of accidentally marking a feed read causes huge data loss and that the confirmation dialogue on folders causes some grief.

I think I could bang this out in a few hours.

**Comment by manderson23** [comment]
Sounds good.

**Comment by samuelclay** [comment]
I disagree that marking a feed as read is data loss. The workflow of a reader is to mark stories as read. And if you accidentally mark a feed as read, you can go back and read it.

I'm not convinced that we should build this and I think this is going to cause issues. If you mark as read on Android and then check the web, it's still unread on the web. That's not a good solution. This needs to be on the backend and I just don't see it happening.

**Comment by dosiecki** [comment]
> If you mark as read on Android and then check the web, it's still unread on the web

I was envisioning the timeout for this being something like 5-10 seconds, tops.  More quickly than one can generally get over to a browser window.

> I disagree that marking a feed as read is data loss.

Might not be the right term.  It is definitely a daily frustration for me that I will accidentally mark all read in a feed where I had been carefully keeping track of my reading progress in a chronological manner.  A quick "oh, crap" button would relieve me from then having to go back into the feed and scroll around to figure out where I was in the list to un-mark it or quickly read all the newer stories.  I will admit that I have very unsteady fingers on my phone, though.

**Comment by samuelclay** [comment]
This is why I want to ship #177. We have this on iOS and it's great. If i do accidentally mark a feed as read, I can go right back into it. I definitely want to ship #177 before this.

**Comment by manderson23** [comment]
My original reason for filing this was nothing to do with accidentally marking as read but was more about getting rid of the mark folder as read confirmation and replacing it with nicer UI.

Being able to undo bulk unreads would be a further benefit. I don't think I've ever done it for a folder/feed but there are a few times where I would have found the undo option useful for mark older/newer read operations.

**Comment by manderson23** [comment]
Another case where undo would be useful
https://getsatisfaction.com/newsblur/topics/android-move-mark-stories-as-read-away-from-save-story-in-menu

**Comment by samuelclay** [comment]
For that I'd prefer to move the Save Story button. Can we move that to below "Send To" and above "Mark as unread"?
